### DISCUSSION TITLE
TypeScript and union

### DISCUSSION THREAD

**Discussion by davetapley** [mobxjs/mobx-state-tree#1906] [Show and tell] [answered]
Following [these docs](https://mobx-state-tree.js.org/tips/typescript#using-a-mst-type-at-design-time) but with a `union`:

```ts
const Foo = types.model({name: types.string, foo: types.string});
const Bar = types.model({name: types.string, bar: types.string});
const FooBar = types.union(Foo, Bar)
interface IFooBar extends Instance<typeof FooBar> {}
```

Yields:
```
An interface can only extend an object type or intersection of object types with statically known members.  TS2312
```
Which I suppose makes sense. Switching to `type` works just fine:

```ts
type FooBar = Instance<typeof FooBar> 
```

Maybe worth mentioning in [these docs](https://mobx-state-tree.js.org/tips/typescript#using-a-mst-type-at-design-time) ?

---

Going further we can give `Foo` and `Bar` a ` type` discriminator

```ts
const Foo = types.model({id: types.identifier, foo: types.string, type: types.literal("foo")});
const Bar = types.model({id: types.identifier, bar: types.number, type: types.literal("bar")});
```

Then we can define a model:
```ts
const Store = types.model({foobars: array(FooBar)})

const store = Store.create({foobars: [
  {id: "1", foo:"test"}, // note omitting type here
  {id: "2", bar: 200}   // and here
]})

store.foobars.map(({type}) => type)
// (2) ['foo', 'bar']
```

⬆️ some how it something automatically inferred `type` 💫 
**How does this magic work?**

---

Finally we can add an `isFoo` type guard and derive `foos`:

```ts
interface IFoo extends Instance<typeof Foo> {}

const isFoo = (foobar: FooBar): foobar is IFoo => foobar.type === "foo";

const Store = types
  .model({foobars: array(FooBar)})
  .views((self) => ({
    get foos() { return self.foobars.filter(isFoo) }
}))
```

Now we have:

```ts
  const strs = store.foos.map(({foo}) => foo) 
```
... `strs` is correctly identified as `string[]`

And:

```ts
store.foos.map(({bar}) => bar)
```

Correctly gives `Property 'bar' does not exist on type 'IFoo'.  TS2339` 🎉 

---

Comments / improvements welcome! 🙏🏻

**Comment by davetapley** [comment, marked as answer]
I edited my original post thus ⬇️, light of #1907

~`const Foo = types.model({id: types.identifier, foo: types.string, type: "foo"});`~
~`const Bar = types.model({id: types.identifier, bar: types.number, type: "bar"});`~

`const Foo = types.model({id: types.identifier, foo: types.string, type: types.literal("foo")});`
`const Bar = types.model({id: types.identifier, bar: types.number, type: types.literal("bar")});`

---

I'm now able to answer my question:
> How does this magic work?

By 'guessing' the first model which _appears_ to match the snapshot:
https://github.com/mobxjs/mobx-state-tree/blob/cf36cfeedc27d6a0877ccb8b8d3e64169924bcee/packages/mobx-state-tree/src/types/utility-types/union.ts#L108-L110

And as I show on #1907 this **will not work if the types differ only by array properties**.

**Reply by coolsoftwaretyler** [reply]
Hey @davetapley - thank you for the excellent work digging through the code on this one. I'm going to mark this as the "answer" since you've answered your own question here, and I'll follow up in #1907 about next steps.
